adapters/unicorn/unicorn.go

```diff
@@ -0,0 +1,251 @@
+				Message: fmt.Sprintf("Error while decoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		var placementID string
+		if ext.Bidder.PlacementID != "" {
+			placementID = ext.Bidder.PlacementID
+		} else {
+			placementID, err = getStoredRequestImpID(imp)
+			if err != nil {
+				return &errortypes.BadInput{
+					Message: fmt.Sprintf("Error get StoredRequestImpID from imp[%d]: %s", i, err),
+				}
+			}
+		}
+
+		ext.Bidder.PlacementID = placementID
+
+		imp.Ext, err = json.Marshal(ext)
+		if err != nil {
+			return &errortypes.BadInput{
+				Message: fmt.Sprintf("Error while encoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		secure := int8(1)
+		imp.Secure = &secure
+		imp.TagID = placementID
+	}
+	return nil
+}
+
+func getStoredRequestImpID(imp *openrtb.Imp) (string, error) {
+	v, err := jsonparser.GetString(imp.Ext, "prebid", "storedrequest", "id")
+
+	if err != nil {
+		return "", fmt.Errorf("stored request id not found: %s", err)
+	}
+
+	return v, nil
+}
+
+func setSourceExt() json.RawMessage {
+	return json.RawMessage(`{"stype": "prebid_server_uncn", "bidder": "unicorn"}`)
+}
+
+func setExt(request *openrtb.BidRequest) (json.RawMessage, error) {
+	accountID, err := jsonparser.GetInt(request.Imp[0].Ext, "bidder", "accountId")
+	if err != nil {
+		accountID = 0
+	}
+	var decodedExt *unicornExt
+	err = json.Unmarshal(request.Ext, &decodedExt)
+	if err != nil {
+		decodedExt = &unicornExt{
+			Prebid: nil,
+		}
+	}
+	decodedExt.AccountID = accountID
+
+	ext, err := json.Marshal(decodedExt)
+	if err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Error while encoding ext, err: %s", err),
+		}
+	}
+	return ext, nil
+}
+
+// MakeBids unpacks the server's response into Bids.
+func (a *adapter) MakeBids(request *openrtb.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected http status code: 400",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected http status code: %d", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			bid := bid
+			var bidType openrtb_ext.BidType
+			for _, imp := range request.Imp {
+				if imp.ID == bid.ImpID {
+					if imp.Banner != nil {
+						bidType = openrtb_ext.BidTypeBanner
+					} else if imp.Native != nil {
+						bidType = openrtb_ext.BidTypeNative
```

Comment:
The `true` branch of this `if` statement is not covered in the tests cases. Could we add a native JSON test that covers it?

Reply:
Native will support it in the future, so I left it, but since it is not currently supported, I will remove this branch.

Reply:
Fixed it

deaf6c0